### PR TITLE
api: Fetch pending state from txpool

### PR DESCRIPTION
**Pull request by stevenroose**
This commit changes the source of the state when asking for the
*pending* state.  Instead of consulting the miner of the node, it uses
teh pending state of the transaction pool.

Two reasons for this change:

1. Introduction of the consensus package + BFT algorithms

With the introduction of the consensus package and the potential BFT
algorithms being built with it, the concept of the "pending block/state"
changes its definition.  In some of these algorithms, pending block can
mean a block that is in the process of validation, but that can no
longer be updated.  This conflicts with the traditional meaning of
"pending state", being the latest block plus all known pending
transactions.

Most notably, the Istanbul implementation has this problem.  This also
results in the values for the pending state being different for a node
that has the miner enabled and the miner disabled.

2. General principle that the miner package should be disabled for
non-mining nodes

So considering the previous point, the use of the miner package for
something else than mining should be evaluated.  I think it makes sense
to try working towards not using it at all when a node does not
participate in the mining/validating process.
Currently, the miner's pending state is used (a.o.) for executing new
incoming transactions (because the EVM needs a "current block" for the
global getters).  I think this should be fairly easy to work around and
this is a first step.

### CODE REVIEW

**Comment by stevenroose** [comment]
This is the related Quorum (Istanbul BFT) issue that made this issue come up: https://github.com/jpmorganchase/quorum/issues/247

**Comment by karalabe** [comment]
> This conflicts with the traditional meaning of
pending block, being the latest block plus all known pending
transactions.

There was never such definition. The pending block was always the next block containing as many transactions as fits is.

Also I don't see what this PR supposed to do. The transaction pool only tracks virtual nonces, but otherwise does not have a state db. You attempt to operate on it as if it was a true state database containing all transactions.

Also the reason we had `StateAndHeaderByNumber` and not `StateByNumber` and `HeaderByNumber` is because your chain can change between invoking the two (and it indeed does for highly active nodes serving lots of requests), so your change introduces a data race.

> General principle that the miner package should be disabled for
non-mining nodes

The miner is the one creating the pending state. Whether your node tries to find PoW is irrelevant, but if it requires the pending state, it needs the same logic as mining itself. Also this part of the code is being reworked currently.

**Comment by stevenroose** [comment]
The rationale of the specific issue can be found here: https://github.com/ethereum/EIPs/issues/650#issuecomment-360085474

Namely, all applications that rely on abigen rely on `PendingNonceAt` (aka `eth_getTransactionCount`) to sign new transactions. For these transactions to be valid, the correct nonce should be returned (i.e. the one taking all pending txs for the account in consideration). The implementation of `eth_getTransactionCount` uses the `StateByNumber` implementation changed in this PR.

> There was never such definition. The pending block was always the next block containing as many transactions as fits is.

Regarding that, let me rephrase. [`PendingNonceAt`](https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go#L401) talks about the "pending state", not really "pending block" as such. I don't know where the concept "pending state" is defined, but I always interpreted that as latest + pending txs.

**Comment by stevenroose** [comment]
The rationale of the specific issue can be found here: https://github.com/ethereum/EIPs/issues/650#issuecomment-360085474

Namely, all applications that rely on abigen rely on `PendingNonceAt` (aka `eth_getTransactionCount`) to sign new transactions. For these transactions to be valid, the correct nonce should be returned (i.e. the one taking all pending txs for the account in consideration). The implementation of `eth_getTransactionCount` uses the `StateByNumber` implementation changed in this PR.

> There was never such definition. The pending block was always the next block containing as many transactions as fits is.

Regarding that, let me rephrase. [`PendingNonceAt`](https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go#L401) talks about the "pending state", not really "pending block" as such. I don't know where the concept "pending state" is defined, but I always interpreted that as latest + pending txs.

I did realize the issue of getting the state and header from different sources. Only one method does that, though: `doCall` which executes tx code and needs to build a header.

I would suggest following this up with a change that totally kills the miner when `--mine` is not set, which I think makes sense. This would mean that we need some source for a new header on top of the latest, but that should not be that hard.

As an alternative to this PR, we could introduce a different endpoint in the API backend that is purely used for accessing "an account's next nonce", i.e. using the txpool for that.

**Comment by stevenroose** [comment]
> The miner is the one creating the pending state.

The miner just takes the pending state from the txpool ([here](https://github.com/ethereum/go-ethereum/blob/master/miner/worker.go#L451)) and filters this.

**Comment by holiman** [comment]
I think the thing 'closest' to what you want, is an additional api-method which would be "give me the nonce that the transaction-pending-pool considers the next available nonce for account x". This could be added to the txpool namespace imo. 

Note, though, that relying on such a method to provide nonces in concurrent requests would be dangerous, since there would always be gaps between getting it and submitting a signed transaction into the pending-pool.

**Comment by stevenroose** [comment]
> I think the thing 'closest' to what you want, is an additional api-method which would be "give me the nonce that the transaction-pending-pool considers the next available nonce for account x". This could be added to the txpool namespace imo.

Considering that abigen and `bind`, both part of this repo, are using `ethclient.PendingNonceAt` exactly for that use case, I assumed that that method was supposed to do what you said. The bind package literally uses that method to decide what nonce to use for its next transaction.

So either the bind package is broken because it uses the pending block instead of the pending state which potentially returns a sequence number that is already used, or either that method is ambiguous like I assumed.

So yes, if a `NextNonceFor(account)` could be part of the txpool namespace API (could be accounts as well if that exists), I'd be glad to contribute as much as I can. Perhaps you could point me to where the "txpool api namespace" lives, I can go from there. I'll also update bind and abigen to use the new method.

**Comment by stevenroose** [comment]
Apparently, there is [`Backend.GetPoolNonce`](https://github.com/ethereum/go-ethereum/blob/762f3a48a00da02fe58063cb6ce8dc2d08821f15/internal/ethapi/backend.go) which is not exposed but could be easily exposed for this purpose.

**Comment by stevenroose** [comment]
If I may quote. This literally confirms how I interpreted PendingNonceAt as opposed to @karalabe's interpretation.

```go
// A PendingStateReader provides access to the pending state, which is the result of all
// known executable transactions which have not yet been included in the blockchain. It is
// commonly used to display the result of ’unconfirmed’ actions (e.g. wallet value
// transfers) initiated by the user. The PendingNonceAt operation is a good way to
// retrieve the next available transaction nonce for a specific account.
type PendingStateReader interface {
	PendingBalanceAt(ctx context.Context, account common.Address) (*big.Int, error)
	PendingStorageAt(ctx context.Context, account common.Address, key common.Hash) ([]byte, error)
	PendingCodeAt(ctx context.Context, account common.Address) ([]byte, error)
	PendingNonceAt(ctx context.Context, account common.Address) (uint64, error)
	PendingTransactionCount(ctx context.Context) (uint, error)
}
```
~ [`PendingStateReader` interface](https://github.com/ethereum/go-ethereum/blob/master/interfaces.go#L185)

```go
// The transaction must be signed and have a valid nonce to be included. Consumers of the
// API can use package accounts to maintain local private keys and need can retrieve the
// next available nonce using PendingNonceAt.
```
~ [`TransactionSender` interface](https://github.com/ethereum/go-ethereum/blob/master/interfaces.go#L170)

**Comment by holiman** [comment]
> or either that method is ambiguous like I assumed.

Oh most definitely. You're not the first to stumble upon that, it's been the cause of many bug reports. 

So, I know there's been talk about removing the pending state in it's current form -- since it's kind of useless for anyone not mining; and most callers actually care about _all_ the pending transactions, not just the ones who happen to be selected for inclusion in a potential next block. 

I agree that the documentation you find is erroneous, and misleading. @karalabe @fjl what would you propose to be the right path forward to make our API less confusing and error-prone?

**Comment by stevenroose** [comment]
I would think a miner-independent "pending state" (as opposed to "pending block") would be ideal. 

All the global variables the EVM exposes for the current block can be "dummied" by just creating a dummy header and use that header for all pending txs. (Difficulty will be the hardest, but can be extracted from some block validation logic, I suppose.)

**Comment by phongtt3** [comment]
Can you pls port this PR to quorum and i can test? Currently i try to modify code in quorum, but build was failed because the param in quorum and go-ethereum is difference

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -583,10 +583,15 @@ type CallArgs struct {
...
 
-	state, header, err := s.b.StateAndHeaderByNumber(ctx, blockNr)
+	state, err := s.b.StateByNumber(ctx, blockNr)
 	if state == nil || err != nil {
 		return nil, 0, false, err
 	}
+	header, err := s.b.HeaderByNumber(ctx, blockNr)
+	if header == nil || err != nil {
+		return nil, 0, false, err
+	}
```
Comment (karalabe) [inline internal/ethapi/api.go]: This is racey. The whole reason behind retrieving the header and the state at the same time is because the pending state can and will change very frequently, often between these two calls (retrieving the state and the header separately). This was our original solution too btw that we needed to fix, took quite a lot to detect and figure out the issue.

File: eth/api_backend.go [hunk 2]
```diff
@@ -81,19 +81,19 @@ func (b *EthApiBackend) BlockByNumber(ctx context.Context, blockNr rpc.BlockNumb
 	return b.eth.blockchain.GetBlockByNumber(uint64(blockNr)), nil
 }
 
-func (b *EthApiBackend) StateAndHeaderByNumber(ctx context.Context, blockNr rpc.BlockNumber) (*state.StateDB, *types.Header, error) {
+func (b *EthApiBackend) StateByNumber(ctx context.Context, blockNr rpc.BlockNumber) (*state.StateDB, error) {
 	// Pending state is only known by the miner
 	if blockNr == rpc.PendingBlockNumber {
-		block, state := b.eth.miner.Pending()
-		return state, block.Header(), nil
+		return b.eth.TxPool().State().StateDB, nil
```
Comment (karalabe) [inline eth/api_backend.go]: AFAIK the transaction pool does not have a meaningful state, it just tracks pending nonces. Won't this break everything?

**Comment by karalabe** [comment]
I don't really think this PR makes sense in it's current form. We updated the nonce retrieval in 1.9.0 to retrieve the nonce from the pool, so that's fixed. The txpool does not have a state otherwise, so this PR doesn't do what you think it does.

**Comment by karalabe** [comment]
The issue is that the txpool's state is just a fake think to track nonces, it doesn't track any real state (balances, contracts, whatnot). I could imagine some consensus engine needs perhaps some specifics, but we should discuss those in the context of relevant PRs, not standalone, because it's hard to follow.

**Comment by stevenroose** [comment]
> I don't really think this PR makes sense in it's current form. We updated the nonce retrieval in 1.9.0 to retrieve the nonce from the pool, so that's fixed. The txpool does not have a state otherwise, so this PR doesn't do what you think it does.

I think that was the major problem we were having at the time, so that seems to be a good fix.
